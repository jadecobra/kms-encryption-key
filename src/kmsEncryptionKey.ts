import {
  ArnPrincipal,
  CompositePrincipal,
  PolicyDocument,
  PolicyStatement,
} from '@aws-cdk/aws-iam';
import { Key, IKey } from '@aws-cdk/aws-kms';
import { Construct, RemovalPolicy, Tags } from '@aws-cdk/core';

export interface IKmsEncryptionKeyProps {
  readonly keyName: string;
  readonly administratorRoleArns?: string[];
  readonly environmentName?: string;
  readonly description?: string;
}

const getArnPrincipals = (administratorRoleArns: string[]): ArnPrincipal[] => {
  return administratorRoleArns.map((arn) => new ArnPrincipal(arn));
};

const getAdministratorArns = (
  administratorRoleArns: string[],
): CompositePrincipal[] => {
  return [new CompositePrincipal(...getArnPrincipals(administratorRoleArns))];
};

const createKeyPolicy = (administratorRoleArns: string[]): PolicyDocument => {
  return new PolicyDocument({
    statements: [
      new PolicyStatement({
        actions: ['kms:*'],
        resources: ['*'],
        principals: getAdministratorArns(administratorRoleArns),
      }),
    ],
  });
};
export class KmsEncryptionKey extends Construct {
  kmsKey: IKey;
  description: string;
  keyName: string;
  environmentName: string;

  constructor(scope: Construct, id: string, props?: IKmsEncryptionKeyProps) {
    super(scope, id);
    this.environmentName = props?.environmentName ?? 'DEV';
    this.description = this.getKeyName(props?.description);
    this.keyName = this.getKeyName(props?.keyName);
    this.kmsKey = this.createKmsKey(props?.administratorRoleArns ?? []);
    this.addTags();
  }

  getKeyName(keyName?: string): string {
    return keyName ?? `${this.environmentName}-kms-encryption-key`;
  }

  createKmsKey(administratorRoleArns: string[]): Key {
    return new Key(this, 'EncryptionKey', {
      alias: this.keyName,
      description: this.description,
      removalPolicy: RemovalPolicy.RETAIN,
      policy: createKeyPolicy(administratorRoleArns),
    });
  }

  addTags() {
    Tags.of(this).add('environment', this.environmentName);
    Tags.of(this).add('alias', this.keyName);
  }
}
