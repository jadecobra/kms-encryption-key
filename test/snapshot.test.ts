import { SynthUtils } from '@aws-cdk/assert';
import { Stack } from '@aws-cdk/core';
import { KmsEncryptionKey } from '../src';

test('KmsEncryptionKey creates KMS Encryption Key with administratorRoleArns', () => {
  const stack = new Stack();
  new KmsEncryptionKey(stack, 'Key', {
    description: 'Test Key',
    kmsAdministratorArns: ['arn:aws:iam::123456789012:role/Administrator'],
    keyName: 'test-key',
  });

  expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
});
