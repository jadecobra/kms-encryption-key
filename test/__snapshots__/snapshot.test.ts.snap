// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KmsEncryptionKey creates KMS Encryption Key with administratorRoleArns 1`] = `
Object {
  "Resources": Object {
    "KeyEncryptionKey5B3DD479": Object {
      "DeletionPolicy": "Retain",
      "Properties": Object {
        "Description": "Test Key",
        "KeyPolicy": Object {
          "Statement": Array [
            Object {
              "Action": "kms:*",
              "Effect": "Allow",
              "Principal": Object {
                "AWS": "arn:aws:iam::123456789012:role/Administrator",
              },
              "Resource": "*",
            },
            Object {
              "Action": Array [
                "kms:Create*",
                "kms:Describe*",
                "kms:Enable*",
                "kms:List*",
                "kms:Put*",
                "kms:Update*",
                "kms:Revoke*",
                "kms:Disable*",
                "kms:Get*",
                "kms:Delete*",
                "kms:ScheduleKeyDeletion",
                "kms:CancelKeyDeletion",
                "kms:GenerateDataKey",
                "kms:TagResource",
                "kms:UntagResource",
              ],
              "Effect": "Allow",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "Tags": Array [
          Object {
            "Key": "alias",
            "Value": "test-key",
          },
          Object {
            "Key": "environment",
            "Value": "DEV",
          },
        ],
      },
      "Type": "AWS::KMS::Key",
      "UpdateReplacePolicy": "Retain",
    },
    "KeyEncryptionKeyAlias217383C0": Object {
      "Properties": Object {
        "AliasName": "alias/test-key",
        "TargetKeyId": Object {
          "Fn::GetAtt": Array [
            "KeyEncryptionKey5B3DD479",
            "Arn",
          ],
        },
      },
      "Type": "AWS::KMS::Alias",
    },
  },
}
`;
